### ISSUE TITLE
archive_entry_xattr_add_entry

### ISSUE THREAD

**Issue by Safari77** [Tarsnap/tarsnap#99]
I don't see the point in the for loop.
And if strdup fails, it leaks mem pointed to by xp.

**Comment by gperciva** [comment]
Thanks for your report about `libarchive/archive_entry.c`, lines 1708-1715.

To be honest, I am completely clueless about that `for` loop:

```
        for (xp = entry->xattr_head; xp != NULL; xp = xp->next)
                ;

        if ((xp = (struct ae_xattr *)malloc(sizeof(struct ae_xattr))) == NULL)
```

It would clearly be valuable to set `xp` to a certain value (by walking through the list) if `xp` was used, but then `xp` is immediately being set to a new value.  Note that this code exists in modern libarchive (in `libarchive/archive_entry_xattr.c`, line 94) as well.

There is indeed a memory leak in strdup in that same function.  I am working with the libarchive developers to improve this code https://github.com/libarchive/libarchive/pull/603 but unfortunately there has been no movement for almost two months.

**Comment by gperciva** [comment]
Thanks for the report!  These were finally accepted in libarchive upstream:
https://github.com/libarchive/libarchive/commit/8135eecf4530796050d0987e825cb6d2f99f6740
https://github.com/libarchive/libarchive/commit/c7fb9682b1a0a98f11c9d62bc238556a6e4f6066
and we have included these fixes in Tarsnap in #163.
